edb/server/protocol/auth_ext/webauthn.py

```diff
@@ -0,0 +1,232 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import dataclasses
+import base64
+import json
+import webauthn
+
+from typing import Any, Optional
+
+from edb.errors import ConstraintViolationError
+from edb.server.protocol import execute
+
+from . import config, data, errors, util, local
+
+
+@dataclasses.dataclass(repr=False)
+class WebAuthnRegistrationChallenge:
+    """
+    Object that represents the ext::auth::WebAuthnRegistrationChallenge type
+    """
+
+    id: str
+    challenge: bytes
+    user_handle: bytes
+    email: str
+
+
+class Client(local.Client):
+    def __init__(self, db: Any):
+        self.db = db
+        self.provider = self._get_provider()
+        self.app_name = self._get_app_name()
+
+    def _get_provider(self) -> config.WebAuthnProvider:
+        provider_name = "builtin::local_webauthn"
+        provider_client_config = util.get_config(
+            self.db, "ext::auth::AuthConfig::providers", frozenset
+        )
+        for cfg in provider_client_config:
+            if cfg.name == provider_name:
+                return config.WebAuthnProvider(
+                    name=cfg.name,
+                    relying_party_origin=cfg.relying_party_origin,
+                    require_verification=cfg.require_verification,
+                )
+
+        raise errors.MissingConfiguration(
+            provider_name, f"Provider is not configured"
+        )
+
+    def _get_app_name(self) -> Optional[str]:
+        return util.maybe_get_config(self.db, "ext::auth::AuthConfig::app_name")
+
+    async def create_registration_options_for_email(self, email: str):
+        registration_options = webauthn.generate_registration_options(
+            rp_id=self.provider.relying_party_id,
+            rp_name=(self.app_name or self.provider.relying_party_origin),
+            user_name=email,
+            user_display_name=email,
+        )
+
+        await self._create_registration_challenge(
+            email=email,
+            challenge=registration_options.challenge,
+            user_handle=registration_options.user.id,
+        )
+
+        return (
+            base64.urlsafe_b64encode(registration_options.user.id).decode(),
+            webauthn.options_to_json(registration_options).encode(),
+        )
+
+    async def _create_registration_challenge(
+        self,
+        email: str,
+        challenge: bytes,
+        user_handle: bytes,
+    ):
+        await execute.parse_execute_json(
+            self.db,
+            """
+with
+    challenge := <str>$challenge,
+    user_handle := <str>$user_handle,
+    email := <str>$email,
+insert ext::auth::WebAuthnRegistrationChallenge {
+    challenge := enc::base64_decode(challenge),
+    user_handle := enc::base64_decode(user_handle),
+    email := email,
+}""",
```

Comment:
I'm fairly sure that with the `execute_json` api you can just pass base64 encoded strings for `bytes` params, and edgedb handles the decoding without you needing to use `enc::base64_decode`, as `execute_json` is the backend for the edgeql over http endpoint. But maybe your way is more explicit?

```suggestion
with
    challenge := <bytes>$challenge,
    user_handle := <bytes>$user_handle,
    email := <str>$email,
insert ext::auth::WebAuthnRegistrationChallenge {
    challenge := challenge,
    user_handle := user_handle,
    email := email,
}""",
```

Reply:
Currently it does not:

<details>
<summary>Server gives me this error:</summary>

```
1. TypeError: Object of type bytes is not JSON serializable

------------------------------ Traceback -------------------------------

    /Users/scotttrinh/edgedb/edgedb/edb/server/protocol/auth_ext/http.py, line 113, in handle_request
        > return await self.handle_webauthn_options(*handler_args)
    /Users/scotttrinh/edgedb/edgedb/edb/server/protocol/auth_ext/http.py, line 835, in handle_webauthn_options
        > await webauthn.create_registration_challenge(
    /Users/scotttrinh/edgedb/edgedb/edb/server/protocol/auth_ext/webauthn.py, line 43, in create_registration_challenge
        > await execute.parse_execute_json(
    edb/server/protocol/execute.pyx, line 523, in parse_execute_json
        > return await execute_json(
    edb/server/protocol/execute.pyx, line 561, in execute_json
        > bind_args = _encode_args(args)
    edb/server/protocol/execute.pyx, line 622, in edb.server.protocol.execute._encode_args
        > jval = _encode_json_value(arg)
    edb/server/protocol/execute.pyx, line 606, in edb.server.protocol.execute._encode_json_value
        > jarg = json.dumps(val, cls=DecimalEncoder)
    /nix/store/263fr5l24468xxbxma197gkafc6v0qxz-python3-3.11.4/lib/python3.11/json/__init__.py, line 238, in dumps
        > **kw).encode(obj)
    edb/server/protocol/execute.pyx, line 602, in edb.server.protocol.execute.DecimalEncoder.encode
        > return super().encode(obj)
    /nix/store/263fr5l24468xxbxma197gkafc6v0qxz-python3-3.11.4/lib/python3.11/json/encoder.py, line 200, in encode
        > chunks = self.iterencode(o, _one_shot=True)
    /nix/store/263fr5l24468xxbxma197gkafc6v0qxz-python3-3.11.4/lib/python3.11/json/encoder.py, line 258, in iterencode
        > return _iterencode(o, 0)
    /nix/store/263fr5l24468xxbxma197gkafc6v0qxz-python3-3.11.4/lib/python3.11/json/encoder.py, line 180, in default
        > raise TypeError(f'Object of type {o.__class__.__name__} '
```

</details>

But @1st1 has suggested moving this serialize/deserialize dance into the execute function directly to avoid needing to do it manually. I'm still hoping there is a way we can avoid needing to serialize it at all and use the binary protocol directly, but maybe I'm thinking about this in the wrong way? Do we always have to serialize to JSON when communicating with the server from the server itself?

Reply:
Oh, actually I think I misunderstood. You're not saying that I need to drop _encoding_, just that passing b64 strings will automatically be decoded on the server. Lemme try that.

Reply:
Yep, looks like that works as you've suggested here, thanks! Still would love the avoid the overhead here, but seems strictly better than what I'm doing here!

Reply:
> But @1st1 has suggested moving this serialize/deserialize dance into the execute function directly to avoid needing to do it manually.

I think that could be done here, in the same way I added support for decimals: https://github.com/edgedb/edgedb/pull/5848/files#diff-32be6e82c8d465d3f9979e75008993c3ad08d677a2456d59c668df53eab631d1

Reply:
Maybe we still want to avoid having to explicitly wrap `b64_encode` to the arguments though.

Reply:
That's what I mean. In that custom JSON encoder, we do b64_encode on any bytes values, so when we call `parse_execute_json`, the params in the query are typed with `<bytes>`, and we can pass bytes in the variables.

Reply:
@jaclarke ⚡ https://github.com/edgedb/edgedb/pull/6791/files/d30b5e6e4cb4e0e29e47b74cd3331e08236334f3..7ad4733f547d553d6bb57fa8113410b792098209